library/core/src/option.rs

```diff
@@ -734,6 +735,128 @@ impl<T> Option<T> {
         }
     }
 
+    const fn get_some_offset() -> isize {
+        if mem::size_of::<Option<T>>() == mem::size_of::<T>() {
+            // niche optimization means the `T` is always stored at the same position as the Option.
+            0
+        } else {
+            assert!(mem::size_of::<Option<T>>() == mem::size_of::<Option<mem::MaybeUninit<T>>>());
+            let some_uninit = Some(mem::MaybeUninit::<T>::uninit());
+            // SAFETY: This gets the byte offset of the `Some(_)` value following the fact that
+            // niche optimization is not active, and thus Option<T> and Option<MaybeUninit<t>> share
+            // the same layout.
+            unsafe {
+                (some_uninit.as_ref().unwrap() as *const mem::MaybeUninit<T>)
+                    .byte_offset_from(&some_uninit as *const Option<mem::MaybeUninit<T>>)
+            }
+        }
+    }
+
+    /// Returns a slice of the contained value, if any. If this is `None`, an
+    /// empty slice is returned. This can be useful to have a single type of
+    /// iterator over an `Option` or slice.
+    ///
+    /// Note: Should you have an `Option<&T>` and wish to get a slice of `T`,
+    /// you can unpack it via `opt.map_or(&[], std::slice::from_ref)`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// #![feature(option_as_slice)]
+    ///
+    /// assert_eq!(
+    ///     [Some(1234).as_slice(), None.as_slice()],
+    ///     [&[1234][..], &[][..]],
+    /// );
+    /// ```
+    ///
+    /// The inverse of this function is (discounting borrowing)
+    /// [`[T]::first`]:
+    ///
+    /// [`[T]::first`]: slice::first
+    ///
+    /// ```rust
+    /// #![feature(option_as_slice)]
+    ///
+    /// for i in [Some(1234_u16), None] {
+    ///     assert_eq!(i.as_ref(), i.as_slice().first());
+    /// }
+    /// ```
+    #[inline]
+    #[must_use]
+    #[unstable(feature = "option_as_slice", issue = "none")]
+    pub fn as_slice(&self) -> &[T] {
+        // SAFETY: This is sound as long as `get_some_offset` returns the
+        // correct offset. Though in the `None` case, the slice may be located
+        // at a pointer pointing into padding, the fact that the slice is
+        // empty, and the padding is at a properly aligned position for a
+        // value of that type makes it sound.
+        unsafe {
+            slice::from_raw_parts(
+                (self as *const Option<T>).wrapping_byte_offset(Self::get_some_offset())
+                    as *const T,
+                self.is_some() as usize,
```

Comment:
Since a big part of this is the optimization potential, would it be worth adding a [codegen test](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#codegen-tests) for at least `Option<usize>` and `Option<NonZeroUsize>`?

Reply:
Thinking more, I do think I'd like to see one.  Here's an untested stab at making one for you to include:
```rust
// compile-flags: -O -Z randomize-layout=no
// only-x86_64
// ignore-debug (the extra assertions get in the way)

#![feature(option_as_slice)]

// CHECK-LABEL: @option_usize_as_slice(
#[no_mangle]
pub unsafe fn option_usize_as_slice(x: &Option<usize>) -> &[usize] {
    // CHECK: %[[ADDR:.+]] = getelementptr inbounds { i64, i64 }, ptr %x, i64 0, i32 1
    // CHECK: %[[LEN:.+]] = load i64, {{ptr|i64\*}} %x
    // CHECK: %[[R0:.+]] = insertvalue { {{ptr|i64\*}}, i64 } undef, {{ptr|i64\*}} %[[ADDR]], 0
    // CHECK: %[[R1:.+]] = insertvalue { {{ptr|i64\*}}, i64 } %[[R0]], i64 %[[LEN]], 1
    // CHECK: ret { {{ptr|i64\*}}, i64 } %[[R1]]
    x.as_slice()
}

// CHECK-LABEL: @option_nonzero_usize_as_slice(
#[no_mangle]
pub unsafe fn option_nonzero_usize_as_slice(x: &Option<std::num::NonZeroUsize>) -> &[std::num::NonZeroUsize] {
    // CHECK: %[[VAL:.+]] = load i64, {{ptr|i64\*}} %x
    // CHECK: %[[NE:.+]] = icmp ne i64 %[[VAL]], 0
    // CHECK: %[[LEN:.+]] = zext i1 %[[NE]] to i64
    // CHECK: %[[R0:.+]] = insertvalue { {{ptr|i64\*}}, i64 } undef, {{ptr|i64\*}} %x, 0
    // CHECK: %[[R1:.+]] = insertvalue { {{ptr|i64\*}}, i64 } %[[R0]], i64 %[[LEN]], 1
    // CHECK: ret { {{ptr|i64\*}}, i64 } %[[R1]]
    x.as_slice()
}
```

Reply:
I think checking that we get no `select` should suffice to prove the optimization, and also be more resilient against LLVM changes.

Reply:
`CHECK-NOT`s aren't necessarily more resilient, since if it just turns the `select` into a `br` or a `switch` then the test will "pass" incorrectly.

But hopefully this will be fine for this test.